developers/ocean-node/node-architecture.md

```diff
@@ -0,0 +1,94 @@
+
+### Architecture Overview
+
+The Node stack is divided into the following layers:
+
+* Network layer (libp2p & HTTP API)
+* Components layer (Indexer, Provider)
+* Modules layer
+
+<figure><img src="../../.gitbook/assets/image (1).png" alt=""><figcaption><p>Ocean Node Infrastructure diagram</p></figcaption></figure>
+
+### Features
+
+* libp2p supports ECDSA key pairs, and node identity should be defined as a public key.
+* Multiple ways of storing URLs:
+  * Choose one node and use that private key to encrypt URLs (enterprise approach).
+  * Choose several nodes, so your files can be accessed even if one node goes down (given at least one node is still alive).
+  * Use MPC to divide the key among multiple nodes (so no one node has the entire key).
+* Protecting URLs during download: use a TEE enclave to decrypt the key (makes sense only for the MPC approach) and read content in the enclave.
+* Supports multiple C2D types:
+  * Light Docker only (for edge nodes).
+  * Ocean C2D (Kubernetes).
+* Each component can be enabled/disabled on startup (e.g., start node without Indexer).
+
+### Nodes and Network Model
+
+Nodes can receive user requests in two ways:
+
+* HTTP API
+* libp2p from another node
+
+They are merged into a common object and passed to the appropriate component.
+
+Nodes should be able to forward requests between them if the local database is missing objects. (Example: Alice wants to get DDO id #123 from Node A. Node A checks its local database. If the DDO is found, it is sent back to Alice. If not, Node A can query the network and retrieve the DDO from another node that has it.)
+
+Nodes' libp2p implementation:
+
+* Should support core protocols (ping, identify, kad-dht for peering, circuit relay for connections).
+* For peer discovery, we should support both mDNS & Kademlia DHT.
+* All Ocean Nodes should subscribe to the topic: OceanProtocol. If any interesting messages are received, each node is going to reply.
+
+### Components & Modules
+
+#### Indexer
+
+An off-chain, multi-chain metadata & chain events cache. It continually monitors the chains for well-known events and caches them (V4 equivalence: Aquarius).
+
+Features:
+
+* Monitors MetadataCreated & MetadataUpdated and stores DDOs.
+* Validates DDOs according to multiple SHACL schemas.
+* Provides proof for valid DDOs.
+* Monitors data token contracts & stores orders.
+* Allows queries for all the above.
+
+#### Provider
+
+* Performs checks on-chain for buyer permissions and payments.
+* Encrypts the URL and metadata during publishing.
+* Decrypts the URL when the dataset is downloaded or a compute job is started.
+* Encrypts/decrypts files before storage/while accessing.
+* Provides access to data assets by streaming data (and never the URL).
+* Provides compute services (connects to multiple C2D engines: light, Ocean C2D, third parties).
+
+### Modules
```

Comment:
Should we mention performance testing and system testing somewhere else? I think it's a good idea to throw few works about them.